tools/cache_creator/cache_info.cpp

```diff
@@ -0,0 +1,270 @@
+                           size_t(headerLength), m_cacheBlob.getBufferSize());
+
+  res.trailingSpaceBeforePrivateBlob = headerLength - vk::VkPipelineCacheHeaderDataSize;
+  return res;
+}
+
+// =====================================================================================================================
+// Finds the start offset of the private PipelineBinaryCache header.
+//
+// @returns : Private header offset on success, error if the cache blob does not have a valid private header
+llvm::Expected<size_t> CacheBlobInfo::getPrivateHeaderOffset() const {
+  auto *publicHeader = reinterpret_cast<const vk::PipelineCacheHeaderData *>(m_cacheBlob.getBufferStart());
+  size_t privateHeaderOffset = publicHeader->headerLength;
+
+  if (privateHeaderOffset < vk::VkPipelineCacheHeaderDataSize)
+    return createBlobError(m_cacheBlob, "Vulkan Pipeline Cache header length less than expected");
+
+  if (privateHeaderOffset + sizeof(vk::PipelineBinaryCachePrivateHeader) > m_cacheBlob.getBufferSize())
+    return createBlobError(m_cacheBlob, "Insufficient buffer size for the Pipeline Binary Cache private header");
+
+  return privateHeaderOffset;
+}
+
+// =====================================================================================================================
+// Reads the private PipelineBinaryCache header.
+//
+// @returns : BinaryCachePrivateHeader info on success, error if the cache blob does not have a valid private header
+llvm::Expected<BinaryCachePrivateHeaderInfo> CacheBlobInfo::readBinaryCachePrivateHeaderInfo() const {
+  auto privateHeaderOffsetOrErr = getPrivateHeaderOffset();
+  if (auto err = privateHeaderOffsetOrErr.takeError())
+    return std::move(err);
+
+  BinaryCachePrivateHeaderInfo res = {};
+  res.privateHeader = reinterpret_cast<const vk::PipelineBinaryCachePrivateHeader *>(m_cacheBlob.getBufferStart() +
+                                                                                     *privateHeaderOffsetOrErr);
+  res.contentBlobSize =
+      m_cacheBlob.getBufferSize() - (*privateHeaderOffsetOrErr + sizeof(vk::PipelineBinaryCachePrivateHeader));
+  return res;
+}
+
+// =====================================================================================================================
+// Finds the start offset of the cache content.
+//
+// @returns : Cache content offset on success, error if the cache blob does not have any valid content (not even empty)
+llvm::Expected<size_t> CacheBlobInfo::getCacheContentOffset() const {
+  auto privateHeaderOffsetOrErr = getPrivateHeaderOffset();
+  if (auto err = privateHeaderOffsetOrErr.takeError())
+    return std::move(err);
+
+  return *privateHeaderOffsetOrErr + sizeof(vk::PipelineBinaryCachePrivateHeader);
+}
+
+// =====================================================================================================================
+// Reads all PipelineBinaryCache entries.
```

Comment:
This does not describe everything that is done. It calculates the MD5 hash for the ELF as well.